### PR TITLE
feat!: move to ESM-only and fix package exports

### PR DESCRIPTION
**Pull request by kwaa**
## Description

`unbuild` outputs `.d.mts` and `.d.cts` at build time, but `unocss` doesn't use it.

This PR updates the export map for most packages to export types correctly.

## References

https://publint.dev/unocss@0.57.7
https://arethetypeswrong.github.io/?p=unocss
https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/docs/problems/FalseCJS.md
https://github.com/microsoft/TypeScript/issues/50466
https://github.com/unjs/unbuild/issues/238

### CODE REVIEW

**Comment by userquin** [comment]
ahaha, I was working also on this on my local, checking your PR @kwaa

**Comment by userquin** [comment]
You should use `d.ts` instead `d.cts` in require types, all `d.cts` files should be removed since we're not using type module, those d.cts files are generated by `unbuild` (if packages were with type module, d.mts should be removed and we should use d.ts in import types).

I'll check `cli` package, I guess we should add type module.

I'll check this PR on my local in a few hours...

**Comment by kwaa** [comment]
> find `/* @export-submodules */` and replace it with `/* @xexport-submodules */`: check `scripts/exports.ts` module

Hmm... I didn't quite get that part. Could you please explain it in detail?

**Comment by userquin** [comment]
> Hmm... I didn't quite get that part. Could you please explain it in detail?

forgot it, don't change it

**Comment by kwaa** [comment]
`"type": "module"` is set :)

I ignored the `vscode` package because it only generates an `index.js` file.

`build` and `test` are the same as before.

**Comment by userquin** [comment]
ok, so we only need to fix astro and postcss default exports in cjs, that's easy: https://github.com/antfu/vite-plugin-inspect/pull/99/files#diff-21462feba05a79008c54d9a3267569b27b8897540ac86e47110e57f25402d224R4-R11

**Comment by userquin** [comment]
@kwaa are you going to fix it?

**Comment by kwaa** [comment]
> @kwaa are you going to fix it?

I'm on it. How do I write a Co-authored-by or code comment?

**Comment by userquin** [comment]
We also need to fix `unocss` package exports: 
imagen

**Comment by userquin** [comment]
We need to fix all cjs files generated by `unbuild`, cjs files generated with `tsup`  should be fine.

You should create a new script in scripts folder and call it after `build` script in root package.json file.

**Comment by kwaa** [comment]
> We also need to fix `unocss` package exports

The script currently looks like this:

```ts
import { readFileSync, writeFileSync } from 'node:fs'
import { resolve } from 'node:path'

function patchCjs(cjsModulePath: string, name: string) {
  const cjsModule = readFileSync(cjsModulePath, 'utf-8')
  writeFileSync(
    cjsModulePath,
    cjsModule.replace(`module.exports = ${name};`, `exports.default = ${name};`),
    { encoding: 'utf-8' },
  )
}

patchCjs(resolve('./packages/astro/dist/index.cjs'), 'UnoCSSAstroIntegration')
patchCjs(resolve('./packages/postcss/dist/index.cjs'), 'unocss')
patchCjs(resolve('./packages/unocss/dist/astro.cjs'), 'UnocssAstroIntegration')
patchCjs(resolve('./packages/unocss/dist/postcss.cjs'), 'postcss__default')
```

It works fine.

But considering that there are a lot of packages that also use the default export, I should fix them as well.

###### `unocss`

image

###### `@unocss/astro`

image

###### `@unocss/postcss`

image

**Comment by userquin** [comment]
yes, the script should fix all cjs files in all packages, `unocss` will include a few of them (and so you don't need to install individual packages if you only need defaults provided by `unocss`)

**Comment by userquin** [comment]
For example, removing browser.dts and including both entries in package.json:

imagen

imagen

**Comment by kwaa** [comment]
https://github.com/unocss/unocss/pull/3380/commits/6c24db9b946eea21843193852d1af8e2073cf4cd should fix most of the packages.

Some packages (e.g. `@unocss/nuxt` and `@unocss/runtime` I can't get to, so I'll skip them here.)

> We also need to add `typesVersions` to all packages exporting submodules, this way we can remove all d.ts files in the root folder: for example, `preset-icons` is failing, dts files should use `export = ...`

Okay, I'll take a look tomorrow.

**Comment by userquin** [comment]
@kwaa thx

**Comment by kwaa** [comment]
> Don't remove dts files I will review tmr

Should I add typesVersions and wildcard exports now, or just wait for review?

**Comment by userquin** [comment]
Adding patch to tsup cjs/d.cts files, for runtime package:

imagen

**Comment by userquin** [comment]
@kwaa fixed a few packages, about vite and svelte-scoped: reverted CJS  to Masquerading as ESM, I'll check Vite package to split entries.

**Comment by userquin** [comment]
uhmm, I need to exclude dts file in scope package, it is just a hack...

**Comment by userquin** [comment]
@kwaa you can give it a try, I cannot review pending packages today...

**Comment by kwaa** [comment]
Ah, seems I forgot to update the exports of `@unocss/vite`, fixed now.

image

**Comment by kwaa** [comment]
@userquin The exports of `@unocss/svelte-scoped` seem to be reverted [here](https://github.com/unocss/unocss/pull/3380/commits/061be2967d8abca2dae08a83a6564698a90c7b10#diff-f8613b08dc104a265696013308ba60eeca1bdbd40091886de0846731568e654c)

### Before
image

### After
image

**Comment by kwaa** [comment]
Fixed `@unocss/inspector`.

image

**Comment by userquin** [comment]
I think we should patch all `d.cts` files, if you check logs when building there are warning from unbuild using `patchDCts` in postbuild package:

imagen

imagen

**Comment by kwaa** [comment]
> I think we should patch all `d.cts` files and maybe `d.ts`, if you check logs when building there are warning from unbuild using `patchDCts` in postbuild package:

I don't know much about this. Could you please patch it?

**Comment by userquin** [comment]
> I don't know much about this. Could you please patch it?

I'll ask Anthony, the PR LGTM, I need to do some test: preset icons in examples with errors

File: scripts/postbuild.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  content += `\nexport = ${name};`
+  writeFileSync(
+    file,
+    content,
+    { encoding: 'utf-8' },
+  )
+}
+
+// @unocss/astro
+patchCjs(resolve('./packages/astro/dist/index.cjs'), 'UnoCSSAstroIntegration')
```
Comment (antfu) [inline scripts/postbuild.ts:42]: I don't feel this is a long term solution, as everytime we change the export we need to remeber to update here. And also there is no way to gurad this when it goes wrong. I guess we should have a more automated way.

**Comment by antfu** [comment]
How about dropping CJS and doing ESM only directly?

I think we need to keep CJS for 

- webpack
- eslint
- postcss

and the rest can be ESM-only and imported dynamically in CJS package

**Comment by kwaa** [comment]
> How about dropping CJS and doing ESM only directly?

Since Vite 5 has deprecated the CJS Node API, I think ESM only is fine.

File: scripts/postbuild.ts (hunk 1 above)
Comment (userquin) [reply scripts/postbuild.ts:42]: yeah, that was first approach, we can extract logic via regex, maybe we should fix this in unbuild

**Comment by userquin** [comment]
@kwaa can you update your branch?

@antfu I guess we should only move to ESM-only Vite, Nuxt and Astro packages, if we're consuming packages using webpack....

**Comment by kwaa** [comment]
> @kwaa can you update your branch?

Unfortunately I was getting ready for bed... maybe tomorrow.

File: scripts/postbuild.ts (hunk 1 above)
Comment (userquin) [reply scripts/postbuild.ts:42]: this is only about the default export (in CJS), the rest of exports can be there

File: packages/astro/package.json [hunk 2]
```diff
@@ -26,17 +26,11 @@
...
     ".": {
-      "import": {
-        "types": "./dist/index.d.ts",
-        "default": "./dist/index.mjs"
-      },
-      "require": {
-        "types": "./dist/index.d.cts",
-        "default": "./dist/index.cjs"
-      }
+      "types": "./dist/index.d.mts",
```
[the remark this answers is not in the document]
Comment (userquin) [reply packages/astro/package.json:29]: d.ts

**Comment by kwaa** [comment]
Should pass the test now.

I think it's also possible to make the `unocss` package ESM only, but it currently includes webpack and postcss integration.

**Comment by kwaa** [comment]
> just keep emit cjs and update the corresponding entries

If there are no other packages that need ESM only, I think this PR should be ready for review.

**Comment by kwaa** [comment]
> uhmm, we should also change the rest of packages to be ESM-only, only `webpack`, `eslint-*`and `postcss` should have dual esm/cjs: I only change `autocomplete`

Okay, it looks like postcss and webpack use separate `uno.config.ts` config files instead of function params.

I'll starting later.

File: packages/inspector/package.json [hunk 3]
```diff
@@ -23,17 +23,11 @@
...
     ".": {
-      "import": {
-        "types": "./dist/index.d.ts",
-        "default": "./dist/index.mjs"
-      },
-      "require": {
-        "types": "./dist/index.d.cts",
-        "default": "./dist/index.cjs"
-      }
+      "types": "./dist/index.d.ts",
```
Comment (userquin) [inline packages/inspector/package.json:26]: should be `d.mts`

Comment (userquin) [reply packages/inspector/package.json:26]: review the rest of packages, should use `d.mts` in package exports

**Comment by userquin** [comment]
`runtime` and `unocss` packages should be moved to ESM-only: `unocss` package should emit cjs and only `webpack`  and `postcss` subpackages should be use dual esm/cjs, the rest should be ESM-only

File: packages/inspector/package.json (hunk 3 above)
Comment (kwaa) [reply packages/inspector/package.json]: > yes

Should I use `.d.mts` or `.d.ts` for `exports.import.types` in dual packages?

current:

```json
{
  "exports": {
    ".": {
      "import": {
        "types": "./dist/index.d.ts",
        "default": "./dist/index.mjs"
      },
      "require": {
        "types": "./dist/index.d.cts",
        "default": "./dist/index.cjs"
      }
    }
  },
  "main": "./dist/index.cjs",
  "module": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
}
```

**Comment by userquin** [comment]
uhmm, no idea why vue-cli4 passing tests on main branch, maybe we need to test with Node16, rn I'm using Node 18

**Comment by kwaa** [comment]
It might be less cumbersome to convert the `@unocss/webpack` and `@unocss/postcss` dependencies back to dual packages.

- `@unocss/config`
- `@unocss/core`
- `@unocss/rule-utils`

**Comment by userquin** [comment]
vue-cli4 cannot use Nde16, I misread the test

**Comment by kwaa** [comment]
Is there anything else we need to do?

**Comment by userquin** [comment]
@kwaa cleanup vue config files in webpack examples

I need to review playground in local, some svg logo not being resolved using base (I guess a bug in vite or vue plugin)

**Comment by userquin** [comment]
@kwaa the svg icon working in the PR preview, I'll ignore the problem in dev. I'm going to cleanup vue config files and update webpack entry in integrations section (I'll push changes) and then I'll do the final review.

**Comment by userquin** [comment]
we also need to fix some examples, rn not working, for example nuxt2 throws an error:

imagen

**Comment by userquin** [comment]
Here my review:

ESM-only: type module and node16 (cjs) via dynamic import
ESM-first: type module with dual esm/cjs package exports

This PR changes the following packages:
- `@unocss/astro`: ESM-only
- `@unocss/autocomplete`: ESM-only
- `@unocss/cli`: ESM-only
- `@unocss/config`: ESM-only
- `@unocss/core`: ESM-only
- `@unocss/eslint-config`: ESM first (`@unocss/eslint-config` and `@unocss/eslint-config/flat` package exports)
- `@unocss/eslint-plugin`: ESM first
- `@unocss/extractor-arbitrary-variants`: ESM-only
- `@unocss/extractor-mdc`: ESM-only
- `@unocss/extractor-pug`: ESM-only
- `@unocss/extractor-svelte`: ESM-only
- `@unocss/inspector`: ESM-only
- `@unocss/nuxt`: ESM-only
- `@unocss/post`: ESM-first
- `@unocss/preset-attributify`: ESM-first
- `@unocss/preset-icons`: ESM-only  (`@unocss/preset-icons`, `@unocss/preset-icons/browser` and `@unocss/preset-icons/core` package exports)
-  `@unocss/preset-mini`: ESM-only  (`@unocss/preset-mini`, `@unocss/preset-mini/theme`, `@unocss/preset-mini/variants`, `@unocss/preset-mini/rules`, `@unocss/preset-mini/colors` and `@unocss/preset-mini/utils` package exports)
- `@unocss/preset-rem-to-px`: ESM-only
- `@unocss/preset-tagify`: ESM-only
- `@unocss/preset-typography`: ESM-only
- `@unocss/preset-uno`: ESM-only  (`@unocss/preset-uno`, `@unocss/preset-uno/theme`, `@unocss/preset-uno/colors` and `@unocss/preset-uno/utils` package exports)
- `@unocss/preset-web-fonts`: ESM-only
- `@unocss/preset-wind`: ESM-only
- `@unocss/reset`: only css (type module)??
- `@unocss/rule-utils`: ESM-only
- `@unocss/runtime`: ESM-only
- `@unocss/scope`: ESM-first
- `@unocss/shared-common`, `@unocss/shared-docs` and  `@unocss/shared-integration`: we don't publish these packages (type module)
- `@unocss/svelte-scoped`: ESM-only  (`@unocss/svelte-scoped/preprocess` and `@unocss/svelte-scoped/vite` package exports)
- `@unocss/transformer-attributify-jsx`: ESM-only
- `@unocss/transformer-attributify-jsx-babel`: ESM-only
- `@unocss/transformer-compile-class`: ESM-only
- `@unocss/transformer-directives`: ESM-only
- `@unocss/transformer-variant-group`: ESM-only
- `unocss`: type module with subpacke export mix (@kwaa great work with unbuild 2.0 ;) )
  - `unocss`: ESM-only
  - `unocss/preset-attributify`: ESM-only
  - `unocss/preset-tagify`: ESM-only
  - `unocss/preset-icons`: ESM-only
  - `unocss/preset-mini`: ESM-only
  - `unocss/preset-typography`: ESM-only
  - `unocss/preset-uno`: ESM-only
  - `unocss/preset-web-fonts`: ESM-only
  - `unocss/preset-wind`: ESM-only
  - `unocss/vite`: ESM-only
  - `unocss/astro`: ESM-only
  - `unocss/webpack`: ESM-first
  - `unocss/postcss`: ESM-first
- `@unocss/vite`: ESM-only  (`@unocss/vite` and `@unocss/vite/client` package exports)
- `@unocss/vscode`: CJS only (not modified)
- `@unocss/webpack`: ESM-first

We need to review and fix examples, for example Nuxt 2 not working and also the changes included in webpack integration docs page.

**Comment by userquin** [comment]
uhmmm, why `packages/preset-icons/src/collections.json` removed?

**Comment by userquin** [comment]
maybe we need to fix CJS packages built with `unbuild`: https://github.com/antfu/vite-plugin-inspect/issues/100

**Comment by kwaa** [comment]
> maybe we need to fix CJS packages built with `unbuild`: [antfu/vite-plugin-inspect#100](https://github.com/antfu/vite-plugin-inspect/issues/100)

Perhaps we could stop using the default export, which would avoid a lot of interoperability issues.

**Comment by userquin** [comment]
> Perhaps we could stop using the default export, which would avoid a lot of interoperability issues.

Yeah, thats another solution, we only need to add `exports.__esModule=true;` in cjs files, dont need to patch default exports in d.ts and d.cts files.

**Comment by antfu** [comment]
Thanks for working on this! I'll defer this to the next minor bump so that it has better separation of feature breaking changes and format breaking changes

**Comment by userquin** [comment]
> Thanks for working on this! I'll defer this to the next minor bump so that it has better separation of feature breaking changes and format breaking changes

maybe v1.0.0 ? ;)

**Comment by userquin** [comment]
@kwaa can you run `pnpm install`  and update lock file?

**Comment by kwaa** [comment]
> @kwaa can you run `pnpm install` and update lock file?

The last few commits have had no dependency updates, and the CI errors should have nothing to do with the lock file.

image

**Comment by userquin** [comment]
@kwaa can you merge main again, node error should be fixed now with #3454

**Comment by kwaa** [comment]
> @kwaa can you merge main again?, node error should be fixed now with #3454

Okay, I'll do this tomorrow.

**Comment by antfu** [comment]
Thanks for the hard work! I'll take over and do the release soon

File: packages/nuxt/index.cjs [hunk 4]
```diff
@@ -0,0 +1,6 @@
+// CommonJS proxy to bypass jiti transforms from nuxt 2 and using native ESM
+module.exports = function (...args) {
```
Comment (userquin) [inline packages/nuxt/index.cjs]: should use `exports.default`, you can run `pnpm pack` and test the tgz here: https://arethetypeswrong.github.io/

mlly will check it

File: packages/nuxt/package.json [hunk 5]
```diff
@@ -25,6 +25,7 @@
   "exports": {
     ".": {
       "types": "./dist/index.d.mts",
+      "require": "./index.cjs",
```
Comment (userquin) [inline packages/nuxt/package.json:28]: we need to review this

Comment (antfu) [reply packages/nuxt/package.json:28]: Feel free to push the fix :)

**Comment by antfu** [comment]
We might need a script to check the dist and make sure there is no `require('unocss')` and `require('unocss/core')` in cjs files.

**Comment by userquin** [comment]
@kwaa I'm going to review latest changes in a few minutes, are you working also on this?

**Comment by kwaa** [comment]
> @kwaa I'm going to review latest changes in a few minutes, are you working also on this?

It doesn't feel like there's much I can do right now, so I just merge updates as needed.

**Comment by userquin** [comment]
Another review (https://github.com/unocss/unocss/pull/3380#issuecomment-1836117226), included this changes:
- `@unocss/nuxt`: ESM first, Node 16 (cjs) Masquerading as ESM (we don't generate CJS)
- `@unocss/postcss`: ESM-first for default package export and ESM-only for `@unocss/postcss/esm`

We only need to update `@unocss/postcss` package.json to fix node 10 types (from @antfu changes), I'll push changes:

imagen

**Comment by kwaa** [comment]
It looks like the new `@unocss/preset-legacy-compat` package makes the CI fail. Maybe it should be converted to ESM only as well?

**Comment by userquin** [comment]
> It looks like the new `@unocss/preset-legacy-compat` package makes the CI fail. Maybe it should be converted to ESM only as well?

checking it

**Comment by userquin** [comment]
@kwaa cannot be ESM-Only, check the name, we need to apply similar build config to Dual in unocss package

**Comment by kwaa** [comment]
> @kwaa cannot be ESM-Only, check the name, we need to apply similar build config to Dual in unocss package

Okay, I'll see what I can do to fix it.

**Comment by userquin** [comment]
> Okay, I'll see what I can do to fix it.

Wait, that package using `@unocss/core`...

**Comment by kwaa** [comment]
> Wait, that package using `@unocss/core`...

Maybe try splitting into `index.ts` and `esm.ts` that imports `@unocss/core`, like `@unocss/postcss`?

**Comment by userquin** [comment]
> Maybe try splitting into `index.ts` and `esm.ts` that imports `@unocss/core`, like `@unocss/postcss`?

I guess we only need to patch the `module.exports` => `exports.default`, the problem is the `dist-verify.ts` module, the cjs is fine since it is using `exports.default` but there is an extra named export.

**Comment by userquin** [comment]
Themodule is exporting `presetLegacyCompat` as named and default.

**Comment by userquin** [comment]
@kwaa undo ESM-Only, remove types from package exports (in the new one preset package) and update `dist-verify.ts` script:

imagen


```ts

  const exportsDefault = 'exports.default'

  let error = false
  await Promise.all(cjsFiles.map(async (file) => {
    const code = await fs.readFile(file, 'utf-8')
    const matches = forbidden.map(r => code.match(r)).filter(Boolean)
    // the CJS module can have exports.default and another module.exports
    // preset-legacy-compat is an example, exporting presetLegacyCompat as default and named
    if (matches.length && !code.match(exportsDefault)) {
      console.error(`\nFound forbidden code in ${file}`)
      console.error(matches.map(i => ` - ${i![0]}`).join('\n'))
      error = true
    }
  }))
``

**Comment by kwaa** [comment]
> @kwaa undo ESM-Only, remove types from package exports (in the new one preset package)

Remove only `exports.['.'] .types` or also remove the root `types`?

**Comment by userquin** [comment]
> Remove only `exports.['.'].types` or also remove the root `types`?

only types entry from package exports

**Comment by kwaa** [comment]
It should be fine now, and it would be great if it would merge before `0.59.0`.

**Comment by userquin** [comment]
We should stop using default package exports in new packages with dual ESM/CJS whenever possible

**Comment by kwaa** [comment]
Maybe it's time to merge it? @userquin @antfu

**Comment by kwaa** [comment]
🥺

**Comment by kwaa** [comment]
🥲

File: docs/integrations/webpack.md [hunk 6]
```diff
@@ -25,18 +26,54 @@ This plugin does not come with any default presets.
   ```
 :::
 
+From UnoCSS version `v0.58.1`, UnoCSS has been moved to ESM-only, and so you need to load your configuration via dynamic import:
```
Comment (antfu) [inline docs/integrations/webpack.md:29]:
```suggestion
From UnoCSS version `v0.59.0`, UnoCSS has been moved to ESM-only, you need to load your configuration via dynamic import:
```

File: docs/integrations/webpack.md [hunk 7]
```diff
@@ -80,3 +117,141 @@ Add `uno.css` to your main entry:
 // main.ts
 import 'uno.css'
 ```
+
+## Frameworks
+
+### Vue + Vue CLI
+
+If you're using [Vue CLI](https://cli.vuejs.org/) with webpack 4/5 with UnoCSS `v0.58.1+`, you need to use the latest [Vue CLI Service](https://cli.vuejs.org/guide/cli-service.html) `v5.0.8` to load your configuration with dynamic import:
```
Comment (antfu) [inline docs/integrations/webpack.md:125]:
```suggestion
If you're using [Vue CLI](https://cli.vuejs.org/) with webpack 4/5 with UnoCSS `v0.59.0`, you need to use the latest [Vue CLI Service](https://cli.vuejs.org/guide/cli-service.html) `v5.0.8` to load your configuration with dynamic import:
```

File: packages/preset-uno/package.json [hunk 8]
```diff
@@ -22,30 +22,26 @@
   "sideEffects": false,
   "exports": {
     ".": {
-      "types": "./dist/index.d.ts",
-      "import": "./dist/index.mjs",
-      "require": "./dist/index.cjs"
+      "types": "./dist/index.d.mts",
+      "import": "./dist/index.mjs"
```
Comment (userquin) [inline packages/preset-uno/package.json:26]: iirc this should be "default" instead "import"

Comment (userquin) [reply packages/preset-uno/package.json:26]: root types should use `.d.ts`, `.d.mts` will use `.js` extension and that's wrong in node10

Comment (userquin) [reply packages/preset-uno/package.json:26]: I'm very busy at work, I'll check the changes this afternoon ( /cc @kwaa )

File: packages/eslint-config/package.json [hunk 9]
```diff
@@ -1,50 +1,50 @@
...
+    "url": "https://github.com/unocss/unocss/issues"
+  },
+  "keywords": [
+    "eslint-config",
+    "eslint"
+  ],
+  "sideEffects": false,
+  "exports": {
+    ".": {
+      "types": "./dist/index.d.ts",
```
Comment (userquin) [inline packages/eslint-config/package.json:24]: remove types here, we're masquerading CJS entry: dual ESM/CJS shouldn't have types

Comment (userquin) [reply packages/eslint-config/package.json:24]: should remove types also in postcss and webpack

**Pull request merged**